Improve bean ordering semantics

mikehearn opened the pull request:
The default order of an object in OrderUtil.getOrder should be zero not LOWEST_PRECEDENCE, as otherwise there is no way to request a bean be last in the list. The intention here is to leave a bean that doesn't request an order un-ordered relative to other beans, so zero is the correct value.

Apply a global ordering to all bean event listeners regardless of requested generic type. Previously, listeners were ordered only relative to other listeners that specified the same generic type, meaning the final ordering could be unexpected in cases where supertypes were matched. Ordering is now done after all the listeners are aggregated.

Fixes #11871

mikehearn committed:
Fix bean event listener orderings and rename BeanDefinitionDelegateSpec to better reflect its actual contents.

The default order of an object in OrderUtil.getOrder should be
zero not LOWEST_PRECEDENCE, as otherwise there is no way to
request a bean be last in the list. The intention here is to
leave a bean that doesn't request an order un-ordered relative
to other beans, so zero is the correct value.

Apply a global ordering to all bean event listeners regardless
of requested generic type. Previously, listeners were ordered
only relative to other listeners that specified the same generic
type, meaning the final ordering could be unexpected in cases
where supertypes were matched. Ordering is now done after all
the listeners are aggregated.

dstepanov commented on inject/src/main/java/io/micronaut/context/DefaultBeanContext.java:
```diff
@@ -1263,27 +1255,31 @@ protected <T> void destroyLifeCycleBean(LifeCycle<?> cycle, BeanDefinition<T> de
-
+                }
+            }
+            OrderUtil.sort(listeners);
```
Why do you need to sort them? They should be sorted once

mikehearn replied:
How do you mean? Listeners can be ordered relative to one another so they must be sorted. It's not enough to sort them within the `Class<?>` key, because there might be two different class keys that both accept the same type and you need to order between those.

dstepanov replied:
Listeners are sorted once, and I don't see a reason to resort them all the time.

mikehearn replied:
The issue is that beans are both created lazily and selected per-type. Each event that's fired can select a different subset of listeners, which might in turn create them (so we can't pre-compute the sorts as the objects don't exist at initialization). In the common case there are zero or one listeners, so we can avoid the sort easily then. 

Do you see a good way to avoid the sort in every case?

dstepanov replied:
What it can create? A listener? By registering it manually?

mikehearn replied:
The listener beans. I'm assuming there's a reason the existing code has this lazy supplier abstraction: the actual bean is created only when there's an event to send it. We could change that to create them all up front but I guess changing the timing of when beans are created is also backwards incompatible.

dstepanov replied:
Creating anything up front would hurt the startup. Not sure about your change, do you need this change for anything?

mikehearn replied:
Yes, we have a listener that records information about beans that might be found in the heap. It must therefore run last, as otherwise the listener will record info about an intermediate bean that might be replaced by a later event listener.

Is this a real concern? A sort of a small array is extremely fast. I can add a branch to skip if .size() <= 1 but I'd be surprised if sorting event listener beans ever came up in profiles unless someone implements `Ordered` in an expensive way.

dstepanov replied:
Please create a test

graemerocher commented on core/src/main/java/io/micronaut/core/order/OrderUtil.java:
```diff
@@ -174,18 +174,19 @@ public static void sort(Object[] objects) {
             return getOrder(metadata);
         }
-        return Ordered.LOWEST_PRECEDENCE;
+        return 0;
```
this is a breaking change that has the potential to break existing applications

mikehearn replied:
Possibly; to break an app bean would have to be trying to give itself a low precedence and failing due to this bug.

What we could do is use a separate comparator to fix the case of bean event listeners specifically, and leave the default ordering alone with a note in the javadoc that it's buggy. It can be marked as deprecated and then things re-unified in Micronaut 5.

dstepanov replied:
Maybe we can leave this method and only change the internal use-cases like listeners.

mikehearn replied:
Yes, I think having a separate comparator and a note to clean it up in 5 is the way to go. I'll do that.

mikehearn commented:
@graemerocher @dstepanov I've added a test, avoided changing the existing OrderUtil API and skip the sort if the list isn't big enough to need one. I also renamed the test file. PTAL.

mikehearn commented:
CI failure looks like a flake rather than a result of this change. I see quite a few flakes locally related to the HTTP/Netty tests.

mikehearn commented:
Everything seems to pass except the flake I already fixed in 4.9.x, so once it's merged up that test would have passed too. I think this is ready for merge.

graemerocher commented:
this needs to target 4.10.x

Change is too high risk to go into 4.9.x this late

mikehearn commented:
@graemerocher It does target 4.10.x now (I rebased and retargeted the PR already).

mikehearn commented:
I tweaked the test. Note that the reason this is so complex is that event listener dispatch works in an odd way - it has a `List<Map.Entry<Class<?>, List<Listener>>>` type for the listener registrations. First we iterate over this outer list and select only the inner lists (map.entry values) that for which the key is assignable from the bean. This looks like it might be an optimization to reduce the number of `Class.isAssignableFrom` calls so I didn't want to remove it, but there's no comment in the code stating that explicitly.

A simpler approach would be to just have a `List<Listener>` and sort it whenever a new listener is added. Then dispatch can be just a loop over all listeners that does an isAssignableFrom test on each element. But I'm not sure if that'd be faster or slower, overall.
